### PR TITLE
feature/CPLP-1170 add external service error handling

### PR DESCRIPTION
**Pull request by qxz2mqe**
https://jira.catena-x.net/browse/CPLP-1170

### CODE REVIEW

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs [hunk 1]
```diff
@@ -402,7 +418,15 @@ await foreach (var companyUser in _portalDBAccess.GetCompanyUserRolesIamUsersAsy
 
         private async Task<bool> DeleteUserInternalAsync(CompanyUser companyUser, string iamIdpAlias)
         {
-            var userIdShared = await _provisioningManager.GetProviderUserIdForCentralUserIdAsync(iamIdpAlias, companyUser.IamUser!.UserEntityId).ConfigureAwait(false);
+            string userIdShared = string.Empty;
+            try
+            {
+                userIdShared = await _provisioningManager.GetProviderUserIdForCentralUserIdAsync(iamIdpAlias, companyUser.IamUser!.UserEntityId).ConfigureAwait(false);
+            }
+            catch (KeycloakEntityNotFoundException ex)
```
Comment (ntruchsess) [inline src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs:426]: this try-catch is redundant, please remove

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ClientManager.cs [hunk 2]
```diff
@@ -53,7 +53,7 @@ public async Task<ClientAuthData> GetCentralClientAuthDataAsync(string internalC
         var credentials = await _CentralIdp.GetClientSecretAsync(_Settings.CentralRealm, internalClientId).ConfigureAwait(false);
         if (credentials == null)
```
Comment (ntruchsess) [inline src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ClientManager.cs:54]: redundant null check, please remove (see previous comment)

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ClientManager.cs [hunk 3]
```diff
@@ -67,7 +67,7 @@ public async Task<ClientAuthData> ResetCentralClientAuthDataAsync(string interna
         var credentials = await _CentralIdp.GenerateClientSecretAsync(_Settings.CentralRealm, internalClientId).ConfigureAwait(false);
         if (credentials == null)
```
Comment (ntruchsess) [inline src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ClientManager.cs:68]: redundant null check, please remove (see previous comment)

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ClientManager.cs [hunk 4]
```diff
@@ -82,7 +82,7 @@ private async Task<Client> GetCentralClientViewableAsync(string clientId)
             .SingleOrDefault();
         if (client == null)
         {
```
Comment (ntruchsess) [inline src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ClientManager.cs:84]: redundant null check, please remove (see previous comment)

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ServiceAccountManager.cs [hunk 5]
```diff
@@ -35,7 +35,7 @@ public async Task<ServiceAccountData> SetupCentralServiceAccountClientAsync(stri
                 serviceAccountUser.Id,
                 await GetCentralClientAuthDataAsync(internalClientId).ConfigureAwait(false));
         }
-        catch(NotFoundException nfe)
+        catch(KeycloakEntityNotFoundException nfe)
```
Comment (ntruchsess) [inline src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ServiceAccountManager.cs:38]: that try-catch is redundant, please remove

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/ProvisioningManager.cs [hunk 6]
```diff
@@ -132,11 +133,20 @@ public async Task<string> SetupClientAsync(string redirectUrl)
...
+            try
             {
-                throw new Exception($"failed to retrieve central user {userId}");
+                user = await _CentralIdp.GetUserAsync(_Settings.CentralRealm, userId).ConfigureAwait(false);
+                if (user == null)
+                {
+                    throw new Exception($"failed to retrieve central user {userId}");
+                }
             }
+            catch (KeycloakEntityNotFoundException ex)
```
Comment (ntruchsess) [inline src/provisioning/CatenaX.NetworkServices.Provisioning.Library/ProvisioningManager.cs:145]: redundant try/catch, please remove

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/ProvisioningManager.cs [hunk 7]
```diff
@@ -149,12 +159,21 @@ public async Task AddBpnAttributetoUserAsync(string userId, IEnumerable<string>
...
             {
-                throw new ArgumentOutOfRangeException($"userId {userId} is not linked to shared realm {realm}");
+                providerUserId = await GetProviderUserIdForCentralUserIdAsync(realm, userId).ConfigureAwait(false);
+                if (providerUserId == null)
+                {
+                    throw new ArgumentOutOfRangeException($"userId {userId} is not linked to shared realm {realm}");
+                }
             }
-            return await _SharedIdp.SendUserUpdateAccountEmailAsync(realm, providerUserId, Enumerable.Repeat("UPDATE_PASSWORD",1)).ConfigureAwait(false);
+            catch (KeycloakEntityNotFoundException ex)
```
Comment (ntruchsess) [inline src/provisioning/CatenaX.NetworkServices.Provisioning.Library/ProvisioningManager.cs:171]: redundant try/catch, please remove and revert the declaration of providerUserId

**Review by ntruchsess — CHANGES_REQUESTED** [review]
As the new EntityNotFoundException shall not be thrown by code that is not releated to keycloak-api-calls I did move the EntityNotFoundException into new namespace CatenaX.NetworkServices.Keycloak.ErrorHandling (creating a new project in the solution) renaming it to KeycloakEntityNotFoundException.
If not being caught by Provisioningmanager or BusinessLogic it will be handled by the unspecific Exception-path of GeneralHttpErrorHandler resulting in Error 500 as expected. I therefore removed the Exception-Type-specific (but redundant) error-handling from GeneralHttpErrorHandler, we just don't need it. If we want something different than error 500 it would be caught and handled in businesslogic (or provisioningManager), and if we want error 500 then we already get it.
Please see my other comments for requested changes (most are about redundant try/catch)

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 8]
```diff
@@ -146,7 +147,14 @@ await foreach (var userData in _applicationRepository.GetInvitedUsersDataByAppli
             if (!userData.BusinessPartnerNumbers.Contains(businessPartnerNumber))
             {
                 userBusinessPartnersRepository.CreateCompanyUserAssignedBusinessPartner(userData.CompanyUserId, businessPartnerNumber);
-                await _provisioningManager.AddBpnAttributetoUserAsync(userData.UserEntityId, Enumerable.Repeat(businessPartnerNumber, 1));
+                try
+                {
+                    await _provisioningManager.AddBpnAttributetoUserAsync(userData.UserEntityId, Enumerable.Repeat(businessPartnerNumber, 1));
+                }
+                catch (KeycloakEntityNotFoundException exception)
```
[the remark this answers is not in the document]
Comment (qxz2mqe) [reply src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs]: fixed

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs [hunk 9]
```diff
@@ -288,8 +289,15 @@ public async Task<int> AddOwnCompanyUsersBusinessPartnerNumbersAsync(Guid compan
             }
 
             var businessPartnerRepository = _portalRepositories.GetInstance<IUserBusinessPartnerRepository>();
-
-            await _provisioningManager.AddBpnAttributetoUserAsync(user.UserEntityId, businessPartnerNumbers).ConfigureAwait(false);
+            try
+            {
+                await _provisioningManager.AddBpnAttributetoUserAsync(user.UserEntityId, businessPartnerNumbers).ConfigureAwait(false);
+            }
+            catch (KeycloakEntityNotFoundException ex)
```
[the remark this answers is not in the document]
Comment (qxz2mqe) [reply src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs]: fixed

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs [hunk 10]
```diff
@@ -324,10 +332,18 @@ public async Task<CompanyUserDetails> UpdateOwnUserDetails(Guid companyUserId, O
...
+            string userIdShared = string.Empty;
+            try
+            {
+                userIdShared = await _provisioningManager.GetProviderUserIdForCentralUserIdAsync(iamIdpAlias, companyUser.IamUser!.UserEntityId).ConfigureAwait(false);
+                if (userIdShared == null)
+                {
+                    throw new NotFoundException($"no shared realm userid found for {companyUser.IamUser!.UserEntityId} in realm {iamIdpAlias}");
+                }
+            }
+            catch (KeycloakEntityNotFoundException ex)
```
[the remark this answers is not in the document]
Comment (qxz2mqe) [reply src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs]: fixed

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs (hunk 1 above)
Comment (qxz2mqe) [reply src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs]: fixed

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs [hunk 11]
```diff
@@ -472,10 +496,18 @@ public async Task<bool> ExecuteOwnCompanyUserPasswordReset(Guid companyUserId, s
...
+                    bool updatedPassword = false;
+                    try
+                    {
+                        updatedPassword = await _provisioningManager.ResetSharedUserPasswordAsync(idpUserName.IdpName, idpUserName.TargetIamUserId).ConfigureAwait(false);
+                        if (!updatedPassword)
+                        {
+                            throw new Exception("password reset failed");
+                        }
+                    }
+                    catch (KeycloakEntityNotFoundException ex)
```
[the remark this answers is not in the document]
Comment (qxz2mqe) [reply src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs]: fixed

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ClientManager.cs (hunk 4 above)
Comment (qxz2mqe) [reply src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ClientManager.cs]: fixed

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ServiceAccountManager.cs (hunk 5 above)
Comment (qxz2mqe) [reply src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/ServiceAccountManager.cs]: fixed

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/UserManager.cs [hunk 12]
```diff
@@ -62,7 +62,7 @@ public async Task<bool> UpdateSharedRealmUserAsync(string realm, string userId,
             var user = await _SharedIdp.GetUserAsync(realm, userId).ConfigureAwait(false);
             if (user == null)
```
[the remark this answers is not in the document]
Comment (qxz2mqe) [reply src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Extensions/UserManager.cs]: fixed

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/ProvisioningManager.cs (hunk 6 above)
Comment (qxz2mqe) [reply src/provisioning/CatenaX.NetworkServices.Provisioning.Library/ProvisioningManager.cs]: fixed

**Pull request merged**